### PR TITLE
[DeviceSanitizer] Support detecting out-of-bounds error on DeviceGlobals

### PR DESCRIPTION
**Pull request by zhaomaosu**
llvm part: https://github.com/AllanZyne/llvm/pull/1

### CODE REVIEW

File: source/loader/layers/sanitizer/asan_interceptor.cpp [hunk 1]
```diff
@@ -36,6 +37,12 @@ constexpr auto kSPIR_DeviceType = "__DeviceType";
 
 constexpr auto kSPIR_DeviceSanitizerReportMem = "__DeviceSanitizerReportMem";
 
+const auto kSPIR_AsanDeviceGlobalNumber = "__AsanDeviceGlobalNumber";
+const auto kSPIR_AsanDeviceGlobalSize = "__AsanDeviceGlobalSize";
+const auto kSPIR_AsanDeviceGlobalSizeWithRedZone =
+    "__AsanDeviceGlobalSizeWithRZ";
+const auto kSPIR_AsanDeviceGlobalAddr = "__AsanDeviceGlobalAddr";
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/asan_interceptor.cpp:44]:
```suggestion
constexpr auto kSPIR_AsanDeviceGlobalNumber = "__AsanDeviceGlobalNumber";
constexpr auto kSPIR_AsanDeviceGlobalSize = "__AsanDeviceGlobalSize";
constexpr auto kSPIR_AsanDeviceGlobalSizeWithRedZone =
    "__AsanDeviceGlobalSizeWithRZ";
constexpr auto kSPIR_AsanDeviceGlobalAddr = "__AsanDeviceGlobalAddr";
```

**Review by AllanZyne — APPROVED** [review]
LGTM

File: source/loader/layers/sanitizer/asan_interceptor.cpp [hunk 2]
```diff
@@ -555,6 +562,80 @@ ur_result_t SanitizerInterceptor::updateShadowMemory(ur_queue_handle_t Queue) {
...
+
+    Devices.resize(PropSize / sizeof(ur_device_handle_t));
+    Result = context.urDdiTable.Program.pfnGetInfo(
+        Program, UR_PROGRAM_INFO_DEVICES, PropSize, Devices.data(), nullptr);
+    if (Result != UR_RESULT_SUCCESS) {
+        context.logger.error(
+            "Failed to get list of devices associated with program: {}",
+            Result);
+        return Result;
+    }
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/asan_interceptor.cpp:587]: NIT, it'd be better to extract it to a function, which will make this part more readable.

Comment (zhaomaosu) [reply source/loader/layers/sanitizer/asan_interceptor.cpp]: Done

File: source/loader/layers/sanitizer/asan_interceptor.cpp [hunk 3]
```diff
@@ -36,6 +37,9 @@ constexpr auto kSPIR_DeviceType = "__DeviceType";
 
 constexpr auto kSPIR_DeviceSanitizerReportMem = "__DeviceSanitizerReportMem";
 
+constexpr auto kSPIR_AsanDeviceGlobalNumber = "__AsanDeviceGlobalNumber";
```
Comment (wenju-he) [inline source/loader/layers/sanitizer/asan_interceptor.cpp:40]:
```suggestion
constexpr auto kSPIR_AsanDeviceGlobalCount = "__AsanDeviceGlobalCount";
```

File: source/loader/layers/sanitizer/asan_interceptor.cpp [hunk 4]
```diff
@@ -173,31 +190,31 @@ ur_result_t SanitizerInterceptor::allocateMemory(
...
         std::scoped_lock<ur_shared_mutex> Guard(ContextInfo->Mutex);
-        ContextInfo->AllocatedUSMMap[AllocBegin] = std::move(AllocInfo);
+        ContextInfo->AllocatedUSMMap[AllocBegin] = std::move(AI);
     }
 
     context.logger.info(
         "AllocInfos(AllocBegin={},  User={}-{}, NeededSize={}, Type={})",
         (void *)AllocBegin, (void *)UserBegin, (void *)UserEnd, NeededSize,
-        Type);
+        (uint32_t)Type);
```
Comment (wenju-he) [inline source/loader/layers/sanitizer/asan_interceptor.cpp:217]:
```suggestion
        static_cast<uint32_t>(Type));
```

Comment (wenju-he) [reply source/loader/layers/sanitizer/asan_interceptor.cpp:217]: perhaps it is better to print a string instead of int. string would be more intuitive

Comment (zhaomaosu) [reply source/loader/layers/sanitizer/asan_interceptor.cpp]: Done

File: source/loader/layers/sanitizer/asan_interceptor.cpp (hunk 3 above)
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/asan_interceptor.cpp]: Done

**Comment by AllanZyne** [comment]
I think you can make PR to UR repo now.

**Pull request merged**